### ISSUE TITLE
Build instructions / cross platform support.

### ISSUE THREAD

**Issue by evanwalpole** [pepelkod/AntBridge#2]
What Linux distro is recommended for compiling on x86 / x64 systems?

Any chance of a port to macOS or Windows?

**Comment by pepelkod** [comment]
Ubuntu 16.04.  I run my in a VM from windows and it works ok.  Just make sure one dongle is attached to the VM and one is going to windows.

My plan is to port it to Raspberrian (sp) and use a raspberry pi zero ($5) to do the work.  Essentially a stand-alone Fortius to Ant dongle.

-Doug
________________________________

**Comment by evanwalpole** [comment]
Okay, was able build and execute in Kali 2017.3.

Had to change in types.h line "#define FALSE ((BOOL) 0)" to "#define FALSE 0" to comply with new ISO C++ standard.

The FE-C trainer device is both transmitting and receiving ERG and SLOPE console commands from external ANT+ apps. However there is no data being passed from the Fortius to the bridge app (no speed, power, cadence number etc. Permanently 0). 

There is about four or five varieties of the Fortius trainer that Tacx produced. What is target head unit you are developing for i.e. T1942? The trainer I am working on has a T1932 head unit, and may have a different I/O stream format.

**Comment by pepelkod** [comment]
1942
https://ultraimg.com/images/2016/11/04/xfRa.jpg

Ok so it appears that the 1932 does not need firmware download.  It should just enumerate as a 1932 unit.  However I don't know how accurate the numbers are going to be.

See, Tacx apparently started sending the 1942 out "dumb" with only a bootloader.  Every time the software started, it loaded the code in the 1942 controller.

I assume they did this for two reasons.  1) No one else could use their hardware.  You had to have software to run the device and the software did not stay on the controller after the power was turned off.  To copy the software was illegal.
2) Every update of the software could have the newest version of the firmware.

The problem is that if they changed, anything in the firmware between versions, I would not support it.  I only currently support the version that I have.

So, to your problem: Look for a VID of 3561, and a PID of 1932.  (lsusb | grep 3561)

Second, make sure usb-serial-simple.ko isn't grabbing the USB port.

Three, modify the LibUsb.ccp line ~283 and and a print "\nconnected!\n"l

Let me know how that works out.

-Doug


________________________________

**Comment by pepelkod** [comment]
Oops. Line 325 is where it checks for your PID.

________________________________

**Comment by pepelkod** [comment]
Evan, did this work for you?  Did you get your 1932 head unit working?

**Comment by robgeurden** [comment]
Hi Douglas,

First of all, great work! I really like my Fortius trainer and it would be a pity if it wasn't usable with Zwift.
I've downloaded an Ubuntu image, loaded it into virtualbox, followed your instructions.  After that I've started the application and it's constantly returning Fortius::run: error, got  length of 64..... 
Also it says power mk 0.0000w, cadence 0.0000 speed 0.000 etc, every twenty lines or so.

When I start up Zwift it does find the ANT+ device, so AntBridge is sending out a signal. Albeit just one with no data. Because Zwift isn't reacting to my biking.

I also got an 1932 head-unit attached, and I couldn't follow the last instructions. Could you clarify what I could try in Ubuntu to get it running? I'm a developer (no C++ btw (C#)), but my knowledge of Linux is really really limited, so it would be cool if you could provide me with some dummy help.

Thanks a lot!
Rob

**Comment by pepelkod** [comment]
Rob,  Unfortunately I don't have a 1932.  I am not sure if Evan got his working but I know someone who did!    Pim Hertogs in this forum I think has it working for the 1932.

https://support.zwift.com/hc/en-us/community/posts/115003812752-Support-for-old-non-ANT-trainers-and-bikes?page=1#community_comment_115001716151

**Comment by robgeurden** [comment]
Hi Douglas,

Thanx for your reply guessing by his name I would say he's also dutch. I will contact him and see how that goes! 

I'll keep you posted if I'll get it working.

Have a good one!

**Comment by robgeurden** [comment]
Hi Douglas,

Reached out to Pim. He's got it working on his Tacx. I did some more debuging and it seems my head unit is sending out 64bytes of data. Your run function only works with 24/48 bytes of data. So that seems to be the root of the issue. 
I don't know the protocol/api of what my head unit is sending out, otherwise I could do some test to see if I would be able to get it working.

Do you perhaps have some more information?

Cheers,
Rob

**Comment by robgeurden** [comment]
Hi Douglas, 

I checked the original sourcde @ https://github.com/GoldenCheetah/GoldenCheetah/blob/5e9a5dc75c27a2f0bc6d092807cd46492e3fac3b/src/Fortius.cpp

And in line 409 it state >= 48, your code only check 24 and 48 bits. So probably this is why it isn't sending/receiving the data. I will modify your code and try again.

I will keep you posted.

**Comment by pepelkod** [comment]
Yes.  Sorry for the delay in answering.  I did change that because I was having read errors on USB through the VM.  I don't know if it fixed it but I think that might work for you.

**Comment by robgeurden** [comment]
Hi Douglas,

No worries, I wanted you to know that changing the if statement fixed the problem. Data is sent to Zwift nicely! 

One last question, should I change/input my weight anywhere in the code? If so in which file/what line should I change this?

Thanks a lot, and great work!

**Comment by pepelkod** [comment]
Yes. Change your weight on line 610 in this file:
https://github.com/pepelkod/AntBridge/blob/master/fortius_code/CANTMaster.cpp

Note you can change bike weight too, but that is pretty insignificant (at least in my case).

Hopefully I can figure out how to get Zwift to send down the weight info, but I have not had time to figure it out yet.  Zwift must send it down to some trainers, as it would make a huge difference but I can't get it to send me the data.

-Douglas Pepelko
________________________________

**Comment by pim555** [comment]
Hi Doug,

In my branch I allow the user and bike weights to be set per commandline. As long as Zwift does not pass weigths, it will take those values into account.

Best
Pim

**Comment by mtbiker22** [comment]
> Hi Douglas,
> 
> No worries, I wanted you to know that changing the if statement fixed the problem. Data is sent to Zwift nicely!
> 
> One last question, should I change/input my weight anywhere in the code? If so in which file/what line should I change this?
> 
> Thanks a lot, and great work!

Hi Rob, 
What did you actually change? I also own a Fortius 1932 and I receive a lot of USB read errors (110)...
Thanks!
